src/commands/update.rs

```diff
@@ -1,11 +1,208 @@
+            anyhow::bail!("Was unable to parse into a package requirement or a package named.")
+        }
+    }
+}
+
+enum DependencyChange {
+    Added(PackageId),
+    Removed(PackageId),
+    Updated { from: PackageId, to: PackageId },
+}
+
+fn generate_depedency_changes(
+    old_dependencies: &BTreeSet<PackageId>,
+    new_dependencies: &BTreeSet<PackageId>,
+) -> Vec<DependencyChange> {
+    let added_packages = new_dependencies.difference(&old_dependencies);
+    let removed_packages = old_dependencies.difference(&new_dependencies);
+    let changed_dependencies: BTreeSet<&PackageName> = added_packages
+        .clone()
+        .chain(removed_packages.clone())
+        .map(|package| package.name())
+        .collect();
+
+    let mut depedency_changes = Vec::new();
+
+    for dependency_name in changed_dependencies {
+        let matching_packages_removed = removed_packages
+            .clone()
+            .filter(|x| *x.name() == *dependency_name);
+        let matching_packages_added = added_packages
+            .clone()
+            .filter(|x| *x.name() == *dependency_name);
+
+        match (
+            matching_packages_added.clone().count(),
+            matching_packages_removed.clone().count(),
+        ) {
+            (1, 1) => {
+                // We know for certain that there is only one item in the iterator.
+                let package_added = matching_packages_added.last().unwrap();
+                let package_removed = matching_packages_removed.last().unwrap();
+                depedency_changes.push(DependencyChange::Updated {
+                    from: package_added.clone(),
+                    to: package_removed.clone(),
+                });
+            }
+            (0, 1) => {
+                // We know for certain that there is only one item in the iterator.
+                let package_removed = matching_packages_removed.last().unwrap();
+                depedency_changes.push(DependencyChange::Removed(package_removed.clone()));
+            }
+            (1, 0) => {
+                // We know for certain that there is only one item in the iterator.
+                let package_added = matching_packages_added.last().unwrap();
+                depedency_changes.push(DependencyChange::Added(package_added.clone()));
+            }
+            (0, 0) => panic!("Impossible for the package name {} to not be removed or added if found in earlier.", dependency_name),
+            (_, _) => {
+                let mut found_changes = matching_packages_added
+                    .map(|package| DependencyChange::Added(package.clone()))
+                    .chain(
+                        matching_packages_removed
+                            .map(|package| DependencyChange::Removed(package.clone())),
+                    )
+                    .collect();
+                depedency_changes.append(&mut found_changes)
+            }
+        }
+    }
+
+    depedency_changes
```

Comment:
```suggestion
    let mut dependency_changes = Vec::new();

    dependency_changes.extend(new_dependencies.difference(old_dependencies).map(
        |new_dependency| {
            match old_dependencies
                .iter()
                .find(|old_dependency| old_dependency.name() == new_dependency.name())
            {
                Some(old_dependency) => DependencyChange::Updated {
                    from: old_dependency.clone(),
                    to: new_dependency.clone(),
                },
                None => DependencyChange::Added(new_dependency.clone()),
            }
        },
    ));

    dependency_changes.extend(
        old_dependencies
            .difference(new_dependencies)
            .filter(|old_dependency| {
                !new_dependencies
                    .iter()
                    .any(|new_dependency| old_dependency.name() == new_dependency.name())
            })
            .map(|old_dependency| DependencyChange::Removed(old_dependency.clone())),
    );

    dependency_changes
```

Reply:
While simpler, it doesn't cover all the edge cases. There are cases where the same package, but different versions, are added or removed.

Reply:
Good point! Looks like I got a bit lost in the sauce while coming up with that change and forgot about that.